### DISCUSSION TITLE
NLOHMANN_DEFINE_TYPE with inheritance?

### DISCUSSION THREAD

**Discussion by mqnc** [nlohmann/json#3322] [Ideas]
Would it be possible to provide macros like NLOHMANN_DEFINE_TYPE_INTRUSIVE that account for inheritance?
Can [this](https://github.com/nlohmann/json/issues/2199) mechanism be wrapped into something like

`NLOHMANN_DEFINE_DERIVED_TYPE_INTRUSIVE(derived, base, field1, field2)`

?

**Comment by nlohmann** [comment]
Good news: this has since been implemented. include/nlohmann/detail/macro_scope.hpp now defines `NLOHMANN_DEFINE_DERIVED_TYPE_INTRUSIVE(Type, BaseType, ...)` (plus a `_WITH_NAMES` variant and non-intrusive/with-default counterparts), which does exactly what was requested here — it calls the base class's `to_json`/`from_json` and then handles the derived-only fields.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
